src/form/df.jl

```diff
@@ -0,0 +1,163 @@
+abstract type AbstractDFForm <: AbstractPowerFormulation end
+
+""
+abstract type SOCDFForm <: AbstractDFForm end
+
+""
+const SOCDFPowerModel = GenericPowerModel{SOCDFForm}
+
+"default SOC constructor"
+SOCDFPowerModel(data::Dict{String,Any}; kwargs...) = GenericPowerModel(data, SOCDFForm; kwargs...)
+
+""
+function variable_branch_current(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_branch_series_current_magnitude_sqr(pm, n; kwargs...)
+end
+
+""
+function variable_branch_flow(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_active_branch_flow(pm, n; kwargs...)
+    variable_reactive_branch_flow(pm, n; kwargs...)
+end
+
+""
+function variable_voltage(pm::GenericPowerModel{T}, n::Int=pm.cnw; kwargs...) where T <: AbstractDFForm
+    variable_voltage_magnitude_sqr(pm, n; kwargs...)
+end
+
+"do nothing, this model does not have complex voltage constraints"
+function constraint_voltage(pm::GenericPowerModel{T}, n::Int) where T <: AbstractDFForm
+end
+
+"""
+Defines branch flow model power flow equations
+"""
+function constraint_flow_losses(pm::GenericPowerModel{T}, n::Int, i, f_bus, t_bus, f_idx, t_idx, r, x, g_sh_fr, g_sh_to, b_sh_fr, b_sh_to, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    p_to = pm.var[:nw][n][:p][t_idx]
+    q_to = pm.var[:nw][n][:q][t_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+    ccm =  pm.var[:nw][n][:ccm][i]
+
+    @constraint(pm.model, p_fr + p_to ==  g_sh_fr*(w_fr/tm^2) + r*ccm +  g_sh_to*w_to)
+    @constraint(pm.model, q_fr + q_to == -b_sh_fr*(w_fr/tm^2) + x*ccm + -b_sh_to*w_to)
+end
+
+
+"""
+Defines voltage drop over a branch, linking from and to side voltage magnitude
+"""
+function constraint_voltage_magnitude_difference(pm::GenericPowerModel{T}, n::Int, i, f_bus, t_bus, f_idx, t_idx, r, x, g_sh_fr, b_sh_fr, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+    ccm =  pm.var[:nw][n][:ccm][i]
+
+    #define series flow expressions to simplify Ohm's law
+    p_fr_s = p_fr - g_sh_fr*(w_fr/tm^2)
+    q_fr_s = q_fr + b_sh_fr*(w_fr/tm^2)
+
+    #KVL over the line:
+    @constraint(pm.model, w_to == (w_fr/tm^2) - 2*(r*p_fr_s + x*q_fr_s) + (r^2 + x^2)*ccm)
+
+end
+
+"""
+Defines relationship between branch (series) power flow, branch (series) current and node voltage magnitude
+"""
+function constraint_branch_current(pm::GenericPowerModel{T}, n::Int, i, f_bus, f_idx, g_sh_fr, b_sh_fr, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    ccm  = pm.var[:nw][n][:ccm][i]
+
+    #define series flow expressions to simplify constraint
+    p_fr_s = p_fr - g_sh_fr*(w_fr/tm^2)
+    q_fr_s = q_fr + b_sh_fr*(w_fr/tm^2)
```

Comment:
Same as above, it's not immediately obvious to me that this is correct. Did you double check?  Especially because in my derivation the shunt value did not occur in this constraint.